purchase_advance_payment/wizard/purchase_advance_payment_wizard.py

```diff
@@ -0,0 +1,128 @@
+    journal_id = fields.Many2one(
+        "account.journal",
+        "Journal",
+        required=True,
+        domain=[("type", "in", ("bank", "cash"))],
+    )
+    journal_currency_id = fields.Many2one(
+        "res.currency",
+        "Journal Currency",
+        store=True,
+        readonly=False,
+        compute="_compute_get_journal_currency",
+    )
+    currency_id = fields.Many2one("res.currency", "Currency", readonly=True)
+    amount_total = fields.Monetary(readonly=True)
+    amount_advance = fields.Monetary(
+        "Amount advanced", required=True, currency_field="journal_currency_id"
+    )
+    date = fields.Date(required=True, default=fields.Date.context_today)
+    currency_amount = fields.Monetary(
+        "Curr. amount", readonly=True, currency_field="currency_id"
+    )
+    payment_ref = fields.Char("Ref.")
+
+    @api.depends("journal_id")
+    def _compute_get_journal_currency(self):
+        for wzd in self:
+            wzd.journal_currency_id = (
+                wzd.journal_id.currency_id.id or self.env.user.company_id.currency_id.id
+            )
+
+    @api.constrains("amount_advance")
+    def check_amount(self):
+        if self.amount_advance <= 0:
+            raise exceptions.ValidationError(_("Amount of advance must be positive."))
+        if self.env.context.get("active_id", False):
+            self.onchange_date()
+            if (
+                float_compare(
+                    self.currency_amount,
+                    self.order_id.amount_residual,
+                    precision_digits=2,
+                )
+                > 0
+            ):
+                raise exceptions.ValidationError(
+                    _("Amount of advance is greater than residual amount on purchase")
+                )
+
+    @api.model
+    def default_get(self, fields_list):
+        res = super().default_get(fields_list)
+        purchase_ids = self.env.context.get("active_ids", [])
+        if not purchase_ids:
+            return res
+        purchase_id = fields.first(purchase_ids)
+        purchase = self.env["purchase.order"].browse(purchase_id)
+        if "amount_total" in fields_list:
+            res.update(
+                {
+                    "order_id": purchase.id,
+                    "amount_total": purchase.amount_residual,
+                    "currency_id": purchase.currency_id.id,
+                }
+            )
+
+        return res
+
+    @api.onchange("journal_id", "date", "amount_advance")
+    def onchange_date(self):
+        if self.journal_currency_id != self.currency_id:
+            amount_advance = self.journal_currency_id._convert(
+                self.amount_advance,
+                self.currency_id,
+                self.order_id.company_id,
+                self.date or fields.Date.today(),
+            )
+        else:
+            amount_advance = self.amount_advance
+        self.currency_amount = amount_advance
```

Comment:
It would be better to make `currency_amount` a computed writeable field instead, don't you think?
Then you would avoid having to call `onchange_date` programatically (ie: line 50)

Reply:
Updated as suggested, thanks